inst/schema/ValidateInputResponse.schema.json

```diff
@@ -1,3 +1,52 @@
 {
-	"type": "string"
+  "definitions": {
+    "input_response": {
+      "properties": {
+        "filename": { "$ref": "FileName.schema.json" },
+        "type": {
+          "type": "string"
+        },
+        "data": {
+          "type": "object"
+        }
+      },
+      "additionalProperties": false,
+      "required": [ "filename", "type", "data" ]
+    },
+
+    "pjnz_response": {
+      "allOf": [
+        { "$ref": "#/definitions/input_response" },
+        { "properties": 
+          {
+            "filename": {},
+            "type": { 
+              "const": "pjnz"
+            },
+            "data": { "$ref": "PjnzResponseData.schema.json" }
+          }
+        }
```

Comment:
Could you add the `required` and `additionalProperties` fields to each of the specific responses also? They don't inherit from the `input_response` unfortunately (or at least the package I'm using the generate typescript interfaces doesn't take them as inherited)

Reply:
That seems like a big problem with the validation library, what is it able to get from the allOf?

Reply:
This isn't for validation, it's for actually generating typescript types/interfaces from the schema. I'm using https://www.npmjs.com/package/json-schema-to-typescript.

Is it definitely the case that `addiitonalProperties` and `required` should be inherited from `allOf`? That doesn't seem intuitively correct to me and I can't see anything here to suggest it's the case https://json-schema.org/understanding-json-schema/reference/combining.html#allof

Reply:
From the point of view of validation it will be the case:

* schema A says it has x, y, z and nothing else
* schema B describes the contents of z
* schema C is `allOf` A and B

if the data misses x, y or z or if it has an element w it will fail validation against A via C ("must be valid against all of the given subschemas")

But from the point of view of generating interfaces, I guess all bets are off.  I see above that you're specifically talking about interface generation

Reply:
In fact on closer read of the spec C *can't* specify `additionalProperties` or `required` as well, because those would only apply to properties declared in C itself https://json-schema.org/understanding-json-schema/reference/combining.html

So this is a bug in the schema generation that can't be worked around by changing the schema, feel free to ignore and I'll try to find another solution.

Reply:
Would something that processed the schemas and do the merging in a way that makes sense make the generator's life easier?

Reply:
Ok, so I'm looking at this again and in fact even though each specific response inherits `allOf` "input_response" they each still re-declare all 3 inherited properties, so actually they could each define "additionalProperties" and "required". 

Also I'm not sure why they re-declare the "filename" property as untyped {} (https://github.com/mrc-ide/hintr/pull/11/files#diff-c52a79711c957ee2a7014047c37bd6c6R37) rather than inheriting the typed filename from "input_response" - I think this is incorrect?

The following would generate a correct typescript type:
```
    "shape_response": {
      "allOf": [
        { "$ref": "#/definitions/input_response" },
        { "properties": {
            "type": { 
              "type": "string",
              "enum": [ "shape" ]
            },
            "data": { "$ref": "ShapefileResponseData.schema.json" }
          },
	  "additionalProperties": false,
	  "required": ["type", "data"]
        }
      ]
    }
```

Reply:
Looks good, I'll fix that.

Reply:
What work is "input_response" actually doing here, given that the specific responses re-declare all 3 of its properties?

Reply:
After redeclaring everything, very little - it was there to make one source of truth for things.  The untyped declaration is fine (for the schema) as it is basically saying it's not the responsibility of this subschema to define the type.  But if that causes problems with the generation you're using then we can add the duplication in, and then it might make sense to remove the shared type